cassandra_migrate/migrator.py

```diff
@@ -384,42 +386,68 @@ def _create_version(self, version, migration):
 
         return version_id
 
-    def _apply_migration(self, version, migration, skip=False):
+    def _apply_cql_migration(self, version, migration):
         """
-        Persist and apply a migration
+        Persist and apply a cql migration
 
         First create an in-progress version entry, apply the script, then
         finalize the entry as succeeded, failed or skipped.
+        """
+
+        self.logger.info('Applying cql migration')
+
+        statements = CqlSplitter.split(migration.content)
+
+        try:
+            if statements:
+                self.logger.info('Executing migration -'
+                                 '{} CQL statement'.format(len(statements)))
+
+            for statement in statements:
+                self.session.execute(statement)
+        except Exception:
+            self.logger.exception('Failed to execute migration')
+            raise FailedMigration(version, migration.name)
+
+    def _apply_python_migration(self, version, migration):
+        """
+        Persist and apply a python migration
+
+        First create an in-progress version entry, apply the script, then
+        finalize the entry as succeeded, failed or skipped.
+        """
+        self.logger.info('Applying python script')
+
+        try:
+            check_call([sys.executable, migration.path])
```

Comment:
I would like to make migrations easier to write by setting up the environment as much as possible. I had in mind either:

a) Have the migration define a function with a fixed name (such as `migrate`) that would be called with the Cassandra `Session` as the first argument.

```python
def migrate(session):
    session.execute("BLAH...")
```

b) Have the migration be a complete script, but provide a module function that can retrieve the current `Session` easily.

```python
from cassandra_migrate.migration import get_current_session

session.execute('...')
```

Otherwise it would be quite annoying to reproduce all the connection options in each script.

What do you think about it?

Reply:
Daniel, I feel we are on the same page.

I, personally, prefer not to be too 'opinionated' when it comes to what 'requirements' a template imposes. My reasoning behind tools like this is too make sure that the execution of the script itself works. For example, the current implementation is very basic. All it does is to execute a python script. I would like to add stdout, stderr coming out of migration scripts to be displayed at execution time.

However, there are always common points in any migration. I do see a valid argument for defining some structure to the script like a `migrate` method so that the developers (even those new to python) can ignore the 'ceremony' of a python script and focus on the actual code. I do also see a valid argument on providing a Cassandra session by default because, believe or not, you may be interested on doing some Cassandra operations :D

Another thing is utility methods like the one that executes a query in your original code `q(...)`. So, wherever in my custom script, if I want to execute a query, all I need to do is call `q('SELECT * FROM keyspace.table')` without worrying about the internals.

But aside from that I don't see any other common intersections. I would leave it pretty much more open to the developer and less opinionated from a template perspective. 

For example, we have a case where we want to extract data from a table using the COPY command from cqlsh; then parse UDTs into regular columns; then execute a schema migration (new table with new columns) using the Cassandra session and last, import that data using either COPY or sstableloader tool or whatever gets the job done more efficiently.

As you can see I am using cqlsh, sstableloader, etc., which in many cases most people won't use.

I'll make those changes to the script template and commit them. Looking forward to more suggestions!

Reply:
I don't mind just giving the script the session, without all the utility methods, and let people do how they please. If you need a programmatic migration you're already a bit out of the most convenient path anyway.

What about the module function way?

```
from cassandra_migrate.migration_script import get_session

session = get_session()

...
```

This way the script can still be "freestanding", but just loaded in the same interpreter as cassandra-migrate so that it can use the same context.

Reply:
Daniel,

I am having a hard time trying to understand exactly what to do. I am assuming that you want to 'load ' the script in memory and inject the Cassandra session object? I am trying to figure it out with python but I haven't been successful so far. 

I will try to get back to this PR later this week or next week. In the meantime we are going to use my fork and use what we have so far. We really need to get this migration going even if we have to hardcore some values.

If you have any suggestions on how to do this, I am more than happy to learn.